### DISCUSSION TITLE
Smart Meter Interface: Switch Modbus 8N1 to 8E1 in script

### DISCUSSION THREAD

**Discussion by christophhhh** [arendst/Tasmota#17283] [Q&A]
Hello,
i have a quite complicated heat meter (Integral-V UltraLite PRO) that I would like read out via it's optical Modbus interface.
To do so, the meter needs a wakeup sequence, followed by a command to request data.
The wakeup sequence needs to be sent as 8N1, the following communication is 8E1. (More information can be found here [https://www.mikrocontroller.net/topic/438972](url)

Is there a way to change the parity mode using the Scripting Language?
Apparently there is a way to change the baudrate `sml(m 0 bd)` but I didn't find anything about changing the parity/meter type definition using Scripting Language.

Could I initialize the same physical Sensor (connected to GPIO1+3) twice? 
one using Modbus8N1 (**+1**,3,m,0,2400,,1) for the wakeup sequence and one using Modbus8E1 (**+2**,3,M,0,2400,,1) ?

Thanks in advance!

**Comment by barbudor** [comment]
As of today the command `sml(  )` doesn't support that. Only baudrate. Of course such feature could be added ...

I don't know if you can create 2 meters on the same pins.
But if it works, I would guess that the 1st one would use the hardware UART and the 2nd use the software UART. As only the hardware UART can do 8E1, I would suggest that you test in that order : meter 1 as `M` and meter 2 as `m`
But that's just a guess. Never tested that myself

**Reply by christophhhh** [reply]
Thanks for your hint, I tried the following:

to get data from the meter I have to do this:
- send 0x55 for 2,2 seconds with 8E1, 2400 baud (wakeup sequence)
- wait 400 ms
- send 0x10 0x40 0xFE  0x3E 0x16 with 8N1, 2400 baud
- wait for the answer

How can I achieve that? Especially send 0x55 for such a long time without interruption?

Here is my script, as I don't see anything in the console (with `sensor53 d1` or `sensor53 d2`) I don't think it is executed properly

`>D
;start, define variables
res=0
scnt=0
wkup=1


>B
;setup sensor
->sensor53 r

>F
;this section is executed every 100 ms
scnt+=1
;increment counter

switch scnt
case 1
for wkup 1 532 1
res=sml(1 1 "55")
next
;set string to send to "55" on meter 2 (8E1) 532 times (wake up sequence)
;this should take 2216ms (as required by meter)

res=sml(1 2)
;change to read mode, send nothing

case 26
;   0,1s (start delay)
; + 2,2s wakeup period 
; + 0,3s waiting time (as required by meter)
;is over, we can send the request for data

res=sml(2 1 "1040FE3E16")
;set string to send to "1040FE3E16" on meter 2 (8N1) (request data)

res=sml(2 2)
;set to read mode

case 100
;after 10s > restart sequence
scnt=0
ends

>M 2
;declare sensors

+1,3,M,0,2400,WAERME,1
;Sensor 1: 8E1

+2,3,m,0,2400,WAERME,1
;Sensor 2: 8N1
`

**Comment by gemu2015** [comment]
i added an option (currently only for esp8266)
to restart hardware serial port. e.g.
sml(-1 1 "9600:8E1")
currently supports only baud rate and parity N,E,O
ignores number of bits and stop bits

you may give it a try

[xsns_53_sml.ino.zip](https://github.com/arendst/Tasmota/files/10164258/xsns_53_sml.ino.zip)

**Reply by christophhhh** [reply]
Thanks for your reply! I'll have to dive into how to get this compiled into a firmware, I have never done this but will give it a try
